### PR TITLE
refactor: box component

### PR DESCRIPTION
**Pull request by supproduction**
refactoring for box component

closes [HRZ-2981](https://spryker.atlassian.net/browse/HRZ-2981)

[HRZ-2981]: https://spryker.atlassian.net/browse/HRZ-2981?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: libs/domain/search/box/index.ts [hunk 1]
```diff
@@ -1 +1,3 @@
-export * from './src';
+export * from './box.component';
+export * from './box.model';
+export * from './styles';
```
Comment (tobi-or-not-tobi) [inline libs/domain/search/box/index.ts:4]: let's also export the schema

File: libs/domain/search/box/box.component.ts [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+  categoriesCount: 5,
+  cmsCount: 0,
+})
+@hydratable(['mouseover', 'focusin'])
+@signalAware()
+export class SearchBoxComponent
+  extends ContentMixin<SearchBoxOptions>(LitElement)
+  implements SearchBoxProperties
+{
+  static styles = [baseStyles, searchboxStyles];
```
Comment (tobi-or-not-tobi) [inline libs/domain/search/box/box.component.ts]:
```suggestion
  static styles = [baseStyles, searchBoxStyles];
```

File: libs/domain/search/box/box.component.ts [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+    super();
+    this.onScroll = debounce(this.onScroll.bind(this), 20);
+    this.onTypeahead = debounce(this.onTypeahead.bind(this), 300);
+    this.onClear = this.onClear.bind(this);
+    this.onClose = this.onClose.bind(this);
+  }
+
+  @signalProperty() query = '';
+
+  @query('oryx-typeahead') typeahead!: TypeaheadComponent;
```
Comment (tobi-or-not-tobi) [inline libs/domain/search/box/box.component.ts]:
```suggestion
  @query('oryx-typeahead') protected typeahead?: TypeaheadComponent;
```

File: libs/domain/search/box/box.component.ts [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+  @query('div[slot="option"] > div') scrollContainer?: HTMLElement;
```
Comment (tobi-or-not-tobi) [inline libs/domain/search/box/box.component.ts]:
```suggestion
  @query('div[slot="option"] > div') protected scrollContainer?: HTMLElement;
```

File: libs/domain/search/box/box.component.ts [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+    this.addEventListener('oryx.clear', this.onClear);
+    this.addEventListener('oryx.close', this.onClose);
+  }
+
+  disconnectedCallback(): void {
+    this.removeEventListener('oryx.clear', this.onClear);
+    this.removeEventListener('oryx.close', this.onClose);
+
+    super.disconnectedCallback();
+  }
```
Comment (tobi-or-not-tobi) [inline libs/domain/search/box/box.component.ts]: seems not needed

File: libs/domain/search/box/box.component.ts [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+        ${when(this.query, () => this.renderControls())}
+      </oryx-typeahead>
+    `;
+  }
+
+  protected renderControls(): TemplateResult {
+    return html`
+      <oryx-button slot="suffix" type="text">
+        <button
+          @click=${() => this.dispatchCustomEvent('oryx.clear')}
```
Comment (tobi-or-not-tobi) [inline libs/domain/search/box/box.component.ts]:
```suggestion
          @click=${this.onClear}
```

File: libs/domain/search/box/box.component.ts [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+          @mousedown=${this.muteMousedown}
+        >
+          ${i18n('search.box.clear')}
+        </button>
+      </oryx-button>
+
+      <oryx-icon-button slot="suffix" size=${Size.Sm}>
+        <button
+          aria-label="Close results"
+          @click=${() => this.dispatchCustomEvent('oryx.close')}
```
Comment (tobi-or-not-tobi) [inline libs/domain/search/box/box.component.ts]:
```suggestion
          @click=${this.onClose}
```

File: libs/domain/search/box/box.component.ts [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+          placeholder=${ifDefined(this.$placeholder())}
+        />
+        ${this.renderSuggestion()}
+        ${when(this.query, () => this.renderControls())}
+      </oryx-typeahead>
+    `;
+  }
+
+  protected renderControls(): TemplateResult {
+    return html`
```
Comment (tobi-or-not-tobi) [inline libs/domain/search/box/box.component.ts]:
```suggestion
  protected renderControls(): TemplateResult | void {
    if(!this.query) return;
    
    return html`
```

File: libs/domain/search/box/box.component.ts [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+      </oryx-icon-button>
+    `;
+  }
+
+  protected renderSuggestion(): TemplateResult {
+    const suggestion = this.$suggestion();
+
+    if (!suggestion) {
+      return html``;
+    }
```
Comment (tobi-or-not-tobi) [inline libs/domain/search/box/box.component.ts]: return void?

File: libs/domain/search/box/box.component.ts [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+
+  protected renderProduct(product: Product): TemplateResult {
+    return html`
+      <oryx-content-link
+        class="product"
+        .options="${{
+          type: SemanticLinkType.Product,
+          id: product.sku,
+          label: product.name,
+        }}"
```
Comment (tobi-or-not-tobi) [inline libs/domain/search/box/box.component.ts]:
```suggestion
        .options=${{
          type: SemanticLinkType.Product,
          id: product.sku,
          label: product.name,
        }}
```

File: libs/domain/search/box/box.component.ts [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+  implements SearchBoxProperties
+{
+  static styles = [baseStyles, searchboxStyles];
+
+  constructor() {
+    super();
+    this.onScroll = debounce(this.onScroll.bind(this), 20);
+    this.onTypeahead = debounce(this.onTypeahead.bind(this), 300);
+    this.onClear = this.onClear.bind(this);
+    this.onClose = this.onClose.bind(this);
```
Comment (tobi-or-not-tobi) [inline libs/domain/search/box/box.component.ts]: do we need this?

File: libs/domain/search/box/box.component.ts [hunk 12]
```diff
@@ -0,0 +1,382 @@
...
+export class SearchBoxComponent
+  extends ContentMixin<SearchBoxOptions>(LitElement)
+  implements SearchBoxProperties
+{
+  static styles = [baseStyles, searchboxStyles];
+
+  constructor() {
+    super();
+    this.onScroll = debounce(this.onScroll.bind(this), 20);
+    this.onTypeahead = debounce(this.onTypeahead.bind(this), 300);
```
Comment (tobi-or-not-tobi) [inline libs/domain/search/box/box.component.ts]: can we add this in a alternative way to not use the constructor?

**Review by tobi-or-not-tobi — CHANGES_REQUESTED** [review]

File: libs/domain/search/box/box.component.ts (hunk 5 above)
Comment (supproduction) [reply libs/domain/search/box/box.component.ts]: I was thinking that we intentionally add `oryx.clear, oryx.close` api. Ok, I'll drop it

**Review by tobi-or-not-tobi — DISMISSED** [review]

**Review by tobi-or-not-tobi — APPROVED** [review]

**Pull request merged**